runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -468,10 +472,18 @@ private StreamingDataflowWorker(
   public static StreamingDataflowWorker fromOptions(DataflowWorkerHarnessOptions options) {
     ConcurrentMap<String, ComputationState> computationMap = new ConcurrentHashMap<>();
     long clientId = clientIdGenerator.nextLong();
+    ChannelCache channelCache =
```

Comment:
can we move the channel cache usage to StreamingEngineClient/GrpcWindmillServer and below?

Seems like unnecessary to plumb from up here and given that we are coordinating invalidation at that level it doesn't seem like we want to expose it here.

Reply:
Do we want to share the StubFactory between GrpcDispatcherClient and StreamingEngineClient? If yes it might be a good idea to create it externally and inject it into both objects.

Reply:
I think it's ok for them to have their own caches internally.  They will not generally overlap in direct path mode, and in non-direct path mode a separate channel for the metadata is ok.

We wouldn't want a shared cache if the StreamingEngineClient for example invalidates a channel that is being used by the GrpcDispatcherClient.

Reply:
done